src/absil/ilwritepdb.fs

```diff
@@ -0,0 +1,634 @@
+      iddMajorVersion = 0x0;                                                    // VersionMajor should be 0
+      iddMinorVersion = 0x0;                                                    // VersionMinor should be 0
+      iddType = 0x2;                                                            // IMAGE_DEBUG_TYPE_CODEVIEW
+      iddData = iddDataBuffer }                                                 // Path name to the pdb file when built
+
+// Document checksum algorithms
+let guidSourceHashMD5 = System.Guid(0x406ea660u, 0x64cfus, 0x4c82us, 0xb6uy, 0xf0uy, 0x42uy, 0xd4uy, 0x81uy, 0x72uy, 0xa7uy, 0x99uy) //406ea660-64cf-4c82-b6f0-42d48172a799
+let hashSizeOfMD5 = 16
+
+// If the FIPS algorithm policy is enabled on the computer (e.g., for US government employees and contractors)
+// then obtaining the MD5 implementation in BCL will throw. 
+// In this case, catch the failure, and not set a checksum. 
+let checkSum (url:string) =
+    try
+        use file = FileSystem.FileStreamReadShim(url)
+        use md5 = System.Security.Cryptography.MD5.Create()
+        let checkSum = md5.ComputeHash(file)
+        Some (guidSourceHashMD5, checkSum)
+    with _ -> None
+
+//------------------------------------------------------------------------------
+// PDB Writer.  The function [WritePdbInfo] abstracts the 
+// imperative calls to the Symbol Writer API.
+//------------------------------------------------------------------------------
+
+let sortMethods showTimes info =
+    reportTime showTimes (sprintf "PDB: Defined %d documents" info.Documents.Length)
+    Array.sortInPlaceBy (fun x -> x.MethToken) info.Methods
+    reportTime showTimes (sprintf "PDB: Sorted %d methods" info.Methods.Length)
+    ()
+
+let GetRowCounts tableRowCounts =
+    let builder = ImmutableArray.CreateBuilder<int>(tableRowCounts |> Array.length)
+    tableRowCounts |> Seq.iter(fun x -> builder.Add(x))
+    builder.MoveToImmutable()
+
+let fixupOverlappingSequencePoints fixupSPs showTimes methods =
+    // This next bit is a workaround.  The sequence points we get 
+    // from F# (which has nothing to do with this module) are actually expression 
+    // marks, i.e. the source ranges they denote are typically 
+    // nested, and each point indicates where the  
+    // code for an expression with a particular range begins.  
+    // This is in many ways a much more convenient form to emit. 
+    // However, it is not the form that debug tools accept nicely. 
+    // However, sequence points are really a non-overlapping, non-nested 
+    // partition of the source code of a method.  So here we shorten the 
+    // length of all sequence point marks so they do not go further than 
+    // the next sequence point in the source. 
+    let spCounts =  methods |> Array.map (fun x -> x.SequencePoints.Length)
+    let allSps = Array.concat (Array.map (fun x -> x.SequencePoints) methods |> Array.toList)
```

Comment:
I'm still trying to figure out what I prefer, this codebase has every possible combination.  In fact looking at mine it is more imperative, I may still have some growing to do.